pkg/parser/format.go

```diff
@@ -0,0 +1,343 @@
+package parser
+
+import (
+	"io"
+	"strings"
+)
+
+func newFormatting() *formatting {
+	return &formatting{
+		wss:         map[*BinaryExpression]bool{},
+		comments:    map[Node]string{},
+		indentLevel: -1,
+	}
+}
+
+type formatting struct {
+	w io.StringWriter
+
+	wss         map[*BinaryExpression]bool
+	comments    map[Node]string
```

Comment:
Here we save comments and whitespace concerns purely with the formatting struct and leave the AST alone. Unfortunately this causes an error with tinygo 0.27.0 which has since been fixed on it's dev branch.